runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java

```diff
@@ -118,14 +120,16 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
             try {
               for (;;) {
                 Optional<Boolean> result = checkForSuccess(job);
-                if (result.isPresent()) {
+                if (result.isPresent() && (!result.get() || atMaxWatermark(job))) {
```

Comment:
should this be:
if (result.isPresent() || atMaxWatermark(job)) {...}

Otherwise, if checkForSuccess returns Optional.absent(), it will not check max watermark.